### PR TITLE
🎅 Detect yml instead of yaml

### PR DESCRIPTION
**Pull request by flip111**
https://github.com/purescript/spago/issues/1136

### CODE REVIEW

File: core/src/Config.purs [hunk 1]
```diff
@@ -481,7 +481,13 @@ readConfig :: forall a. FilePath -> Spago (LogEnv a) (Either String { doc :: Yam
 readConfig path = do
   logDebug $ "Reading config from " <> path
   FS.exists path >>= case _ of
-    false -> pure $ Left $ case path of
-      "spago.yaml" -> "Did not find " <> path <> " Run `spago init` to initialise a new project."
-      _ -> "Did not find " <> path
+    false -> case path of
+      "spago.yaml" -> do
+        hasYml <- FS.exists "spago.yml"
```
Comment (klntsky) [inline core/src/Config.purs:486]: It would be better to search for `spago.yml` in the same directory where the config file (`path`) is located.

File: core/src/Config.purs [hunk 2]
```diff
@@ -481,7 +481,13 @@ readConfig :: forall a. FilePath -> Spago (LogEnv a) (Either String { doc :: Yam
...
+        pure $ Left $
+          if hasYml then
+            "Spago's default configuration file must be called spago.yaml The file spago.yml was found, rename it if you will, or run `spago init` to initialise a new project."
+          else
+            "Did not find spago.yaml Run `spago init` to initialise a new project."
+      _ -> pure $ Left $ "Did not find " <> path
```
Comment (klntsky) [inline core/src/Config.purs:492]: These two cases (lines 491-492) are functionally the same. Use pattern matching on two variables or `case` with guards to combine them

File: core/src/Config.purs [hunk 3]
```diff
@@ -481,7 +481,13 @@ readConfig :: forall a. FilePath -> Spago (LogEnv a) (Either String { doc :: Yam
...
   FS.exists path >>= case _ of
-    false -> pure $ Left $ case path of
-      "spago.yaml" -> "Did not find " <> path <> " Run `spago init` to initialise a new project."
-      _ -> "Did not find " <> path
+    false -> case path of
+      "spago.yaml" -> do
+        hasYml <- FS.exists "spago.yml"
+        pure $ Left $
+          if hasYml then
+            "Spago's default configuration file must be called spago.yaml The file spago.yml was found, rename it if you will, or run `spago init` to initialise a new project."
```
Comment (klntsky) [inline core/src/Config.purs:489]:
```suggestion
            "Spago's default configuration file must be called spago.yaml, not spago.yml. Rename it or run `spago init` to initialize a new project."
```

**Review by f-f — COMMENTED** [review]
We should add a few tests with fixtures that verify that we are returning the right error messages at the right time

File: core/src/Config.purs (hunk 3 above)
Comment (f-f) [reply core/src/Config.purs:489]: Spago tends to use the British spellings (`initialise`) over the american ones 😄

File: core/src/Config.purs (hunk 1 above)
Comment (f-f) [reply core/src/Config.purs:486]: Indeed, we use this function also when looking for workspace packages, where it's possible that the user has used the `.yml` extension.

I would use [`String.stripSuffix`](https://pursuit.purescript.org/packages/purescript-strings/6.0.1/docs/Data.String#v:stripSuffix) to detect if the config we're looking for has a `.yml` extension, instead of matching only on the root config

File: src/Spago/Config.purs [hunk 4]
```diff
@@ -155,7 +155,7 @@ readWorkspace { maybeSelectedPackage, pureBuild } = do
   -- First try to read the config in the root. It _has_ to contain a workspace
   -- configuration, or we fail early.
   { workspace, package: maybePackage, workspaceDoc } <- Core.readConfig "spago.yaml" >>= case _ of
-    Left err -> die [ "Couldn't parse Spago config, error:\n  " <> err, "The configuration file help can be found here https://github.com/purescript/spago#the-configuration-file" ]
+    Left err -> die [ "Couldn't parse Spago config, error:\n  " <> err, ". The configuration file help can be found here https://github.com/purescript/spago#the-configuration-file" ]
```
Comment (f-f) [inline src/Spago/Config.purs:158]: Why this change? The new string will go in a new line, so this is going to look weird. Let's add a test with a fixture to verify how it looks like

Comment (flip111) [reply src/Spago/Config.purs:158]: This should be reverted

File: core/src/Config.purs [hunk 5]
```diff
@@ -481,7 +483,21 @@ readConfig :: forall a. FilePath -> Spago (LogEnv a) (Either String { doc :: Yam
 readConfig path = do
   logDebug $ "Reading config from " <> path
   FS.exists path >>= case _ of
-    false -> pure $ Left $ case path of
-      "spago.yaml" -> "Did not find " <> path <> " Run `spago init` to initialise a new project."
-      _ -> "Did not find " <> path
+    false -> do
+      yml <- case map (_ <> ".yml") $ String.stripSuffix (String.Pattern ".yaml") path of
+        Nothing -> pure Nothing
+        Just y -> do
```
Comment (klntsky) [inline core/src/Config.purs:489]: `traverse`

File: core/src/Config.purs [hunk 6]
```diff
@@ -481,7 +483,21 @@ readConfig :: forall a. FilePath -> Spago (LogEnv a) (Either String { doc :: Yam
...
+          hasYml <- FS.exists y
+          pure $
+            if hasYml then
+              Just y
+            else
+              Nothing
```
Comment (klntsky) [inline core/src/Config.purs:495]:
```suggestion
            pure $ guard hasYml *> pure y
```

File: core/src/Config.purs [hunk 7]
```diff
@@ -481,7 +483,21 @@ readConfig :: forall a. FilePath -> Spago (LogEnv a) (Either String { doc :: Yam
...
+      let
+        message = case yml of
+          Nothing -> "Did not find " <> path
+          Just y -> "Did not find " <> path <> ". Spago's configuration files should end with .yaml. Try renaming " <> y
+      pure $ Left $ case path of
+        "spago.yaml" -> message <> "Run `spago init` to initialize a new project."
```
Comment (klntsky) [inline core/src/Config.purs:501]:
```suggestion
        "spago.yaml" -> message <> " Run `spago init` to initialize a new project."
```

File: core/src/Config.purs (hunk 6 above)
Comment (f-f) [reply core/src/Config.purs:495]: I find `guard` harder to read than the if/else above, let's keep that one

**Comment by JordanMartinez** [comment]
CI may fail because I'm not sure what other tests my be affected by these 'pretty print' changes.

File: core/src/Log.purs [hunk 8]
```diff
@@ -98,10 +98,14 @@ instance Loggable String where
...
-instance Loggable a => Loggable (Array a) where
+instance Loggable (Array Docc) where
+  toDoc = Log.lines
+else instance Loggable a => Loggable (Array a) where
   toDoc = Log.lines <<< map toDoc
 
-instance Loggable a => Loggable (NonEmptyArray a) where
+instance Loggable (NonEmptyArray Docc) where
+  toDoc = Log.lines <<< NEA.toArray
+else instance Loggable a => Loggable (NonEmptyArray a) where
```
Comment (f-f) [inline core/src/Log.purs:108]: Why these instance chains? They seem unnecessary

File: src/Spago/Command/Fetch.purs [hunk 9]
```diff
@@ -377,7 +378,12 @@ getPackageDependencies packageName package = case package of
       Right { yaml: { package: Just { dependencies: (Dependencies deps) } } } -> do
         pure (Just (map (fromMaybe Config.widestRange) deps))
       Right _ -> die [ "Read valid configuration from " <> configLocation, "However, there was no `package` section to be read." ]
-      Left err -> die [ "Could not read config at " <> configLocation, "Error: " <> err ]
+      Left errLines -> die
+        [ Log.text $ "Could not read config at " <> configLocation
+        , Log.text "Error: "
+        -- type annotation needed here to clarify that this is `Doc GraphicsParam`, not `Doc a`
+        , Log.indent $ Log.lines $ map Log.text errLines :: Docc
+        ]
```
Comment (f-f) [inline src/Spago/Command/Fetch.purs:386]: The rest of the codebase uses `toDoc` rather than things like `Dodo.text`, I think it works better for inference?

Also note that we are re-exporting `indent` from the Prelude, so most of the times you don't need to import Dodo at all

File: core/src/Config.purs [hunk 10]
```diff
@@ -479,25 +479,30 @@ legacyPackageSetEntryCodec = CA.object "LegacyPackageSetEntry"
...
+        "spago.yaml", Nothing -> 
+          [ "Did not find `" <> path <> "`. Run `spago init` to initialize a new project." ]
+        "spago.yaml", Just y ->
+          [ "Did not find `" <> path <> "`. Spago's configuration files must end with `.yaml`, not `.yml`. "
+          , "Try renaming `" <> y  <> "` to `" <> path <> "` or run `spago init` to initialize a new project."
+          ]
+        _, Nothing ->
+          ["Did not find `" <> path <> "`."]
+        _, Just y -> 
+            [ "Did not find `" <> path <> "`. Spago's configuration files must end with `.yaml`, not `.yml`. "
```
Comment (flip111) [inline core/src/Config.purs]: This duplicates code, previously it was possible to have only one source line with this message

File: src/Spago/Command/Fetch.purs (hunk 9 above)
Comment (JordanMartinez) [reply src/Spago/Command/Fetch.purs]: Ah, that's what I needed to do. Thanks. This change was largely just to fix the inference stuff.

File: core/src/Log.purs (hunk 8 above)
Comment (JordanMartinez) [reply core/src/Log.purs]: It just removes an unnecessary `map` when we know what the `a` is. I've removed them.

File: core/src/Config.purs (hunk 10 above)
Comment (JordanMartinez) [reply core/src/Config.purs]: Correct. But...

The message (per your original code) was outputting:
> Try renaming spago.yml Run `spago init` to initialize a new project.

I thought it would be better to put an `or` in between those two sentences when the context made sense. In some cases, one is likely the correct choice. And rather than write something weird like:
```
  fold $ Array.catMaybes
    [ Just ....
    , if cond1 then Just "Try renaming spago.yaml to..." else Nothing
    , if cond2 then Just "Run spago init" else Nothing
    , if cond3 then Just "Try ... or run ..." else Nothing
    ]
```
I thought it would be more readable to just duplicate the code. This isn't a performance-critical part of the codebase, so I don't think the duplication is harmful.

Comment (f-f) [reply core/src/Config.purs]: Yeah I agree that this change is good

File: src/Spago/Config.purs [hunk 11]
```diff
@@ -188,7 +193,7 @@ readWorkspace { maybeSelectedPackage, pureBuild } = do
       -- We try to figure out if this package has tests - look for test sources
       hasTests <- FS.exists (Path.concat [ Path.dirname path, "test" ])
       pure $ case maybeConfig of
-        Left e -> Left $ "Could not read config at path " <> path <> "\nError was: " <> e
+        Left eLines -> Left $ "Could not read config at path " <> path <> "\nError was: " <> Array.intercalate "\n" eLines
```
Comment (f-f) [inline src/Spago/Config.purs:196]: The `intercalate` here is a bit of a smell because it unnecessarily cares about formatting the error message returned from above - we should return a `Docc` instead of a `Array String` from the `readConfig` function

**Review by f-f — COMMENTED** [review]
One small thing then this is good to go

**Review by f-f — APPROVED** [review]
